### PR TITLE
Generate inbound/outbound for HTTP/2

### LINKED ISSUE #997: Error when using HTTP/2 protocol on Inbound
### Summary

An error `mismatched route reference and resource lengths: map[outbound:backend:true] != 0"` is seen in the dataplane when HTTP/2 protocol is used in the inbound.

### Steps To Reproduce

1. Start `kuma-cp` in standalone mode
2. Apply the supplied `dp-frontend.yaml` and `dp-backend.yaml`
3. Start a single `kuma-dp`
4. Observe the repeating error in the logs when envoy config is re-conciliated

### Additional Details & Logs

- Version 0.7.1
- Error logs
```
2020-08-28T11:24:13.264Z        ERROR   xds-server.reconcile    inconsistent snapshot   {"snapshot": {"Resources":[{"Version":"","Items":{"backend":{"cluster_name":"backend","endpoints":[{"lb_endpoints":[{"HostIdentifier":{"Endpoint":{"address":{"Address":{"SocketAddress":{"address":"10.128.0.54","PortSpecifier":{"PortValue":9000}}}}}},"metadata":{"filter_metadata":{"envoy.lb":{"fields":{"kuma.io/protocol":{"Kind":{"StringValue":"http2"}}}},"envoy.transport_socket_match":{"fields":{"kuma.io/protocol":{"Kind":{"StringValue":"http2"}}}}}},"load_balancing_weight":{"value":1}}]}]}}},{"Version":"","Items":{"backend":{"name":"backend","ClusterDiscoveryType":{"Type":3},"eds_cluster_config":{"eds_config":{"ConfigSourceSpecifier":{"Ads":{}}}},"connect_timeout":{"seconds":5},"http2_protocol_options":{},"LbConfig":null}}},{"Version":"","Items":{}},{"Version":"","Items":{"outbound:10.128.0.15:8000":{"name":"outbound:10.128.0.15:8000","address":{"Address":{"SocketAddress":{"address":"10.128.0.15","PortSpecifier":{"PortValue":8000}}}},"filter_chains":[{"filters":[{"name":"envoy.http_connection_manager","ConfigType":{"TypedConfig":{"type_url":"type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager","value":"EgNtaXAqDgoMZW52b3kucm91dGVyGhIKAhoAEgxvdXRib3VuZDptaXA="}}}]}],"traffic_direction":2}}},{"Version":"","Items":{}},{"Version":"","Items":null}]}, "proxyError": "json: unsupported type: xds.TrafficPermissionMap", "error": "mismatched route reference and resource lengths: map[outbound:backend:true] != 0"}
```
- Configuration
`dp-frontend.yaml`
```yaml
type: Dataplane
mesh: default
name: dp-frontend
networking:
  address: 127.0.0.1
  inbound:
  - port: 6000
    tags:
      kuma.io/service: frontend
      kuma.io/protocol: tcp
  outbound:
  - port: 9000
    address: 127.0.0.1
    tags:
      kuma.io/service: backend
```

`dp-backend.yaml`
```yaml
type: Dataplane
mesh: default
name: dp-backend
networking:
  address: 127.0.0.1
  inbound:
  - port: 9000
    servicePort: 9001
    tags:
      kuma.io/service: backend
      kuma.io/protocol: http2
```

- Platform and Operating System Linux/amd64

reported by @LukeCz

### PR DESCRIPTION
**Pull request by nickolaev**
### Summary

As noted in #997, we have gaps in handling HTTP/2 and gRPC when generating the XDS configuration Snapshots. This PR proposes a fix for this on all the relevant places, plus adds/fixes the relevant tests.

### Issues resolved

Fix #997 #991 

### Documentation

- fixing internal issue

### CODE REVIEW

File: pkg/xds/generator/inbound_proxy_generator.go [hunk 1]
```diff
@@ -53,8 +53,10 @@ func (g InboundProxyGenerator) Generate(ctx xds_context.Context, proxy *model.Pr
 		filterChainBuilder := func() *envoy_listeners.FilterChainBuilder {
 			filterChainBuilder := envoy_listeners.NewFilterChainBuilder()
 			switch protocol {
+			// configuration for HTTP case
 			case mesh_core.ProtocolHTTP:
```
Comment (lobkovilya) [inline pkg/xds/generator/inbound_proxy_generator.go:57]: I'd still prefer :) 
```suggestion
			case mesh_core.ProtocolHTTP, mesh_core.ProtocolHTTP2:
```

File: pkg/xds/generator/outbound_proxy_generator.go [hunk 2]
```diff
@@ -58,8 +58,13 @@ func (g OutboundProxyGenerator) Generate(ctx xds_context.Context, proxy *model.P
...
 
-		// Generate route, routes are only applicable to the HTTP
-		if protocol == mesh_core.ProtocolHTTP {
+		// Generate route, routes are only applicable to the HTTP, HTTP/2 and GRPC protocols
+		switch protocol {
+		case mesh_core.ProtocolHTTP:
+			fallthrough
+		case mesh_core.ProtocolHTTP2:
+			fallthrough
+		case mesh_core.ProtocolGRPC:
```
Comment (lobkovilya) [inline pkg/xds/generator/outbound_proxy_generator.go:67]: and
```suggestion
		case mesh_core.ProtocolHTTP, mesh_core.ProtocolHTTP2, mesh_core.ProtocolGRPC:
```

File: pkg/xds/generator/outbound_proxy_generator.go [hunk 3]
```diff
@@ -101,9 +106,9 @@ func (_ OutboundProxyGenerator) generateLDS(proxy *model.Proxy, subsets []envoy_
 				Configure(envoy_listeners.HttpAccessLog(meshName, envoy_listeners.TrafficDirectionOutbound, sourceService, serviceName, proxy.Logs[serviceName], proxy)).
 				Configure(envoy_listeners.HttpOutboundRoute(envoy_names.GetOutboundRouteName(serviceName))).
 				Configure(envoy_listeners.GrpcStats())
-		case mesh_core.ProtocolHTTP2:
-			fallthrough
 		case mesh_core.ProtocolHTTP:
+			fallthrough
+		case mesh_core.ProtocolHTTP2:
```
Comment (lobkovilya) [inline pkg/xds/generator/outbound_proxy_generator.go:111]:
```suggestion
		case mesh_core.ProtocolHTTP, mesh_core.ProtocolHTTP2:
```

**Review by lobkovilya — APPROVED** [review]
I know this is a matter of taste, I've simply put here my suggestions

**Comment by tharun208** [comment]
@nickolaev #991 is also fixed by this PR i think.

**Comment by jakubdyszkiewicz** [comment]
Double tested with HTTP/2 and gRPC server client

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**